### PR TITLE
Custom ports

### PR DESCRIPTION
**Pull request by dvd-z**
@hannah-bulmer and I worked on https://github.com/elasticio/sftp-component/issues/14.

Some work is based off of Zain and Shanthosh's branch, PR here: https://github.com/elasticio/sftp-component/pull/23

### CODE REVIEW

File: lib/attachments.js [hunk 1]
```diff
@@ -2,71 +2,57 @@ const client = require('elasticio-rest-node')();
...
-    return Promise.resolve();
+  opts.method = 'PUT';
+  opts.headers = {
+    'Content-Length': contentLength,
+  };
+  return opts;
 }
 
-
 function uploadFile(stream, options) {
```
Comment (ShkarupaNick) [inline lib/attachments.js]: Maybe, have the sense use AttachmentProcessofor downloading/uploading attachments.

File: lib/actions/upload.js [hunk 2]
```diff
@@ -1,79 +1,65 @@
...
+  if (!isExists) {
+    await sftp.mkdir(dir, true);
+  }
+  // eslint-disable-next-line no-restricted-syntax
+  for (const key of Object.keys(msg.attachments)) {
+    const attachment = msg.attachments[key];
+    const cur = await sftp.cwd();
+    const targetPath = (cur.charAt(0) === '/') ? path.posix.resolve(dir, key) : path.resolve(dir, key);
+
+    await sftp.put((await request({
```
[the remark this answers is not in the document]
Comment (hannah-bulmer) [reply lib/actions/upload.js]: @ShkarupaNick I cannot find any information about what `responseType: 'stream'` means, would you be able to elaborate? I was looking for it [here](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/responseType)

File: verifyCredentials.js [hunk 3]
```diff
@@ -1,21 +1,16 @@
-var sftp = require("./lib/sftp.js");
-
-module.exports = verify;
+const sftp = require('./lib/sftp.js');
 
 function verify(cfg, cb) {
```
Comment (ElBabbler) [inline verifyCredentials.js]: Better to leave some logs here

Comment (dvd-z) [reply verifyCredentials.js]: Just in `verifyCredentials` or throughout?

**Comment by dvd-z** [comment]
@hannah-bulmer and I refactored things to comply more with ES6 standards

File: lib/actions/upload.js [hunk 4]
```diff
@@ -1,79 +1,47 @@
-'use strict';
-const Client = require('ssh2-sftp-client');
-const request = require('request');
+/* eslint-disable no-await-in-loop */
+const { Logger } = require('@elastic.io/component-commons-library');
```
Comment (uaArsen) [inline lib/actions/upload.js]: In latest version of Sailor you can use this.logger

File: lib/actions/upload.js [hunk 5]
```diff
@@ -1,79 +1,47 @@
...
-}
-
-/**
- * cfg - This is the same config as the one passed to "processMessage" method of the trigger or action
- * startData - result from the startup
- */
-module.exports.shutdown = function shutdownHook() {
-    sftp.end();
+exports.process = async function processAction(msg, cfg) {
+  const logger = Logger.getLogger();
```
Comment (uaArsen) [inline lib/actions/upload.js]: this.logger

File: lib/attachments.js [hunk 6]
```diff
@@ -1,72 +1,16 @@
...
-        const opts = createRequestOptions(contentLength, result.put_url);
-
-        return uploadFile(stream, opts)
-            .then(addUrlAttachment.bind(null, msg, name, result.get_url, contentLength));
-    }
+const { AttachmentProcessor } = require('@elastic.io/component-commons-library');
+
+async function addAttachment(msg, name, stream, contentLength) {
+  // eslint-disable-next-line no-return-await
+  return await new AttachmentProcessor().uploadAttachment(stream, 'stream')
```
Comment (uaArsen) [inline lib/attachments.js]: Refactor this to try/catch with await logic. Also remove `eslint-disable-next-line no-return-await`

File: lib/triggers/read.js [hunk 7]
```diff
@@ -1,278 +1,76 @@
...
-
-    var path = sftp.createPath(dir, PROCESSED_FOLDER_NAME);
-
-    console.log("'%s' does not exist. About to mkdir.", path);
-
-    client.mkdir(path, {mode: 16877}, cb);
-
-}
+exports.process = async function process(msg, cfg) {
+  const logger = Logger.getLogger();
```
Comment (uaArsen) [inline lib/triggers/read.js]: this.logger

File: package.json [hunk 8]
```diff
@@ -1,10 +1,11 @@
 {
   "name": "sftp-component",
-  "version": "0.0.1",
+  "version": "1.1.0",
   "description": "elastic.io SFTP component",
   "scripts": {
-    "test": "NODE_ENV=test ./node_modules/jasmine-node/bin/jasmine-node  --verbose ./spec/*",
-    "integration-test": "NODE_ENV=test mocha --exit spec-integration/*"
+    "pretest": "eslint spec spec-integration lib verifyCredentials.js --fix",
```
Comment (ElBabbler) [inline package.json]: `verifyCredentials.js` is not utit test - it's integration, please check it

File: lib/actions/upload.js [hunk 9]
```diff
@@ -1,79 +1,47 @@
-'use strict';
-const Client = require('ssh2-sftp-client');
-const request = require('request');
+/* eslint-disable no-await-in-loop */
+const { Logger } = require('@elastic.io/component-commons-library');
+const { promisify } = require('util');
+const request = promisify(require('request'));
```
Comment (ElBabbler) [inline lib/actions/upload.js]: Better to upload attachments with `component-commons-library` imstead of `request`

File: lib/actions/upload.js [hunk 10]
```diff
@@ -1,79 +1,42 @@
...
+    await sftp.put(file.data, targetPath, { encoding: null });
+
+    result.results.push({
+      attachment: key,
+      uploadedOn: new Date().toISOString(),
+      fileName: targetPath,
+    });
+  }
+  await this.emit('data', eioUtils.newMessageWithBody(result));
+  await sftp.end();
```
Comment (uaArsen) [inline lib/actions/upload.js]: You can just `return eioUtils.newMessageWithBody(result);`  It is the same to emit('data') + emit('end')

**Pull request merged**